wordpress-mobile/WordPress-Android
# Issue/9739 domain contact form refactoring

This PR is a refactoring of a domain contact form.

I extracted the majority of the logic into ViewModel and added tests.
It's my first time working on ViewModel this big so I would appreciate insight on.. anything, really!
I have been staring on this code for too long, so there is a good chance I missed something.

The PR is giant, and I'm not really sure how to make it smaller since there is a lot of things going behind the scene. Good thing half of it are tests!


To test:

Install the wasabi build.

1. Switch to a site on a business plan with unclaimed domain credit (let me know if you need help with this).
2. Navigate to Plugins screen, and from there to details of any plugin (like Woocommerce).
3. Tap on Install button.
4. In the dialog tap on "REGISTER DOMAIN".
5. Pick a domain (preferably .blog).
6. Tap on "CHOOSE DOMAIN" and navigate to contact form details.
7. If you had contact information stored at WordPress.com it should be prefilled.
8. All the fields apart from State and Organisation are required, and their emptiness is validated locally.
9. The validity of input fields content is validated on backend one by one (eg. if you have two fields with malformed you will have to make two registration attempts to get errors for both of them).
10. Make sure ToS link is clickable.
11. Every time you change the country States are loaded from remote (some countries, like Germany, do not have states).
12. Try to register the domain.
13. PS: Follow domain testing guideline and make sure if you purchase a domain to cancel it.
